### DISCUSSION TITLE
Exposing settings_acceleration_override as a core M-Code

### DISCUSSION THREAD

**Discussion by Dietz0r** [grblHAL/core#412] [Ideas]
Hey,
tinkering with the jerk settings for the motion planner sparked a discussion about using different machining profiles for different machining operations. 
I know Datron has a 5 Step setup for different acceleration and jerk settings and Haas has 3 Profiles they defined as [Roughing, SemiFinish and Finish ](https://www.haas.co.uk/resources/better-techniques/control-features/smooth-operator/). 
I've successfully implemented a [PostProcessor feature](https://media.discordapp.net/attachments/741001353899409449/1187471204332863561/image.png) that simulates that behaviour with changing $120/121/122 on the fly in the [g-code file](https://media.discordapp.net/attachments/741001353899409449/1187495420088307762/image.png)

While sharing that with the community, I was told that there is a [settings_acceleration_override](https://github.com/grblHAL/core/blob/1486280b5e83ec1d2b5c5fb10fa25d55f0ffab82/settings.c#L841) function in grblhal that's only partially exposed in the openpnp plugin (No Z adjustments).  And while this would be a much preferred solution since it's temporary overrides instead of changing controller settings directly i dont want to add mandatory plugins as a dependency for the grblHAL Postprocessor to keep it as general use for every grblHAL user out there as possible.

But i think this might warrant a core M-Code for grblhal since it would open up a lot of options to leverage different kinematics or programs for improved results and is found in lots of commercial controllers.

So could we please get this? :D

**Comment by terjeio** [comment]
In principle I am not too keen on adding "non-standard" gcodes to the core but may make an exception for this.
Is there a defacto standard such a gcode? From the link above it seems Haas uses G187.

> ... found in lots of commercial controllers

Are these using a common M-code or is each using their own?

**Reply by Dietz0r** [reply]
> In principle I am not too keen on adding "non-standard" gcodes to the core but may make an exception for this.

I understand complelty, that's why i'm asking first. :) 

> Is there a defacto standard such a gcode? From the link above it seems Haas uses G187.

I believe that every control manufacturer has their own homebrew for this.
I'll fetch a few more examples of either custom g-codes or straight up variables pushed into a g-code program
ISG-Stuttgart: [Acceleration weighting (G130/G131/G231/G333/G334)](https://www.isg-stuttgart.de/fileadmin/kernel/kernel-html/en-GB/index.html#206590731)
Siemens Sinumeric: [Programmable acceleration override (ACC) (option)](https://support.industry.siemens.com/cs/mdm/98692791?c=14222936331&lc=en-AE)

Together with the Datron and Haas example I believe, that all of those settings are proprietary to the respective control and there's no standard to it but it's a rather common feature in more modern controls.

edit: [found the fanuc way ](https://www.linkedin.com/pulse/fanuc-ai-high-speed-modes-simplified-tim-markoski)

**Reply by Dietz0r** [reply]
I looked a bit more into the Haas implementation.
They have 2 Keywords. The first is for the Servo Profile P1 / P2 / P3 (Rough, Semi, Finish) and an optional E Value.

The E-Value is particularly interesting right now, since i've stared at the junction deviation code for a good while now. The HAAS G187 E-Value sets the junction deviation that the CNC holds the path at. What i found odd is that they run at[ 0.025" (or 0.625mm) per default](https://youtu.be/lwujoXvgGzA?t=220) and tune it down with the g187 commands as needed. [Example: Engraving with E0.05mm](https://youtu.be/cN2jjq9VzvE?t=162)
This just seems massive for me since i'm used to the grblHAL standard value of 0.01mm. 

Maybe there is some worth in including the junction deviation in a maybe planned temporary override function to really speed things up for roughing toolpathes? Though for the record i'd vote to not do profiles but to give it straight up wanted acceleration values. And maybe looking at the default value of that deviation and increasing it a bit? 

PS: Hope you had great holidays with family, friends and some great food! :)

**Reply by terjeio** [reply]
The Fanuc way clashes with LinuxCNC G5.1 used for quadratic splines...
How does Haas G187,E<n> relate to LinuxCNC [G64](http://linuxcnc.org/docs/2.6/html/gcode/gcode.html#sec:G64)? Same idea and can be used instead? Some time back I started looking into implementing G64 and there is a bit of code related to that in the source already (only in the parser and data structures).

Holidays were great except for the weather - alternating rain and snow.., I hope you had a great holiday as well.

**Reply by Dietz0r** [reply]
>How does Haas G187,E relate to LinuxCNC [G64](http://linuxcnc.org/docs/2.6/html/gcode/gcode.html#sec:G64)?

Yep. That seems to be overlapping. HAAS only has G64 as "G64 code cancels exact stop (G61)." and the LinuxCNC P value is exactly what the E-Value of HAAS G187 does. Seems like they rolled the different acceleration settings and path smoothing tolerance into one code.
So we could keep that part to G64 to keep it consistent with LinuxCNC and then just use g187 for giving out acceleration values?

>Holidays were great except for the weather - alternating rain and snow.., I hope you had a great holiday as well.

atleast you had snow T_T ... only rain here. but family time was great and got some nice presents. (Toolholders :D )

**Reply by terjeio** [reply]
I have now added a [config symbol for enabling G61/G64](https://github.com/grblHAL/core/blob/57243a0fc201fe4fdc97ad5880711a19eea49c0d/config.h#L192) in the parser. I am unsure about the [G187 P](https://www.haascnc.com/service/codes-settings.type=gcode.machine=mill.value=G187.html) values, Are they to be taken literally or used for looking up preconfigured values? Anyway, IMO G64 and G187 should not be enabled at the same time, perhaps the config symbol value should be used to distinguish between them - 1 for G64 and 2 for G187? And the symbol should be used to guard your new code and thus make it a compile time option?

Here is a [G64 post](https://www.forum.linuxcnc.org/20-g-code/44022-understanding-g64-behavior) that may be helpful. Complicated stuff this...

**Reply by Dietz0r** [reply]
>I am unsure about the G187 P values, Are they to be taken literally or used for looking up preconfigured values?

The Haas implementation only specifies 3 profiles. The corresponding acceleration and jerk values are stored in control parameters.
Since grblHAL runs on a multitude of different machines with very different settings this will be tedious to setup by the user and add many settings while also being restrictive on how many profiles you can use. 

I would suggest a simple Faktor for the grblHAL implemenation? (If you want to use the G187 code)
G187  P0.4 -> temporary use 40% of the max acceleration ($120,$121,$122etc.) settings. 
G187  P0.6 -> temporary use 60% of the max acceleration settings.
Either disable with P1.0 or a plain G187? 

This would minimize additional settings and gives great situational control for that feature. 
If you wanted to, you could also add an axis word to adjust on an axis basis but, to me, that looks like feature creep already.

>IMO G64 and G187 should not be enabled at the same time, perhaps the config symbol value should be used to distinguish between them - 1 for G64 and 2 for G187? And the symbol should be used to guard your new code and thus make it a compile time option?

On this point I wholeheartedly disagree. 
Best results in terms of machining time and surface finishes are a factor of both these settings. 
Since the machine can use higher accelerations and travel along a wider path if the junction_deviation is higher for maximum roughing time savings or tune down the accelerations and get a really narrow junction_deviation to hit those feature right on the programmed path for finishing. 
As we are using round tools we're only starting to run into issues if the junction_deviation is higher than half the tool diameter. but this is a CAM check unless the machine is aware of what kind of tool is loaded, which it rarely is.

>Here is a [G64 post](https://www.forum.linuxcnc.org/20-g-code/44022-understanding-g64-behavior) that may be helpful. Complicated stuff this...

Yeah it really is not easy, so i'm a bit miffed that this drifted into this direction, i didnt want to burden you with this when asking for a way to adjust acceleration parameters. 

It got a bit out of hand, sorry for that.

But for Pathblending grblHAL is nearly there. [It already calculates the radius of the circle arc needed for a smoothed path. ](https://onehossshay.wordpress.com/2011/09/24/improving_grbl_cornering_algorithm/)
I'll see if i can figure out a way to inject that arc between buffer blocks as its own bufferblock while truncating the vectors of the bufferblocks it gets put in between. Then the junction_deviation setting would work exactly to allow for smooth corner rounding.

Edit2: Or maybe even highjack the input gcode stream and add g2/3 arcs in between those straight lines and then let the normal bufferplanner do it's thing? Would that even be possible? Edit even more: thought more about this, while tempting g2/3 arcs need a plane defined... so we'd lock us out of 3D Z-Motion or couldnt path blend that which is suboptimal... =(

Edit: maybe i can learn something from the LinuxCNC way of pathblending when i can find the right files in the repo. 

The NaiveCam stuff is black magic voodoo to me though. 😅

**Reply by terjeio** [reply]
> I would suggest a simple Faktor for the grblHAL implemenation? (If you want to use the G187 code)
G187 P0.4 -> temporary use 40% of the max acceleration ($120,$121,$122etc.) settings.
G187 P0.6 -> temporary use 60% of the max acceleration settings.
Either disable with P1.0 or a plain G187?

G187 should not be modified this way - a lookup table containing the factors should be used whether constant or not. Or a different gcode if not using the same syntax.

>>    IMO G64 and G187 should not be enabled at the same time, perhaps the config symbol value should be used to distinguish between them - 1 for G64 and 2 for G187? And the symbol should be used to guard your new code and thus make it a compile time option?

> On this point I wholeheartedly disagree.

All of them? IMO at least the code changes should be a compile time option initially. And G64 and G187 belongs to different modal groups (G13 and G0) creating a at least teoretical conflict.

Inserting the arc is not as easy as it seems, the calculation is at the end of adding a line segment and will most likely be done when the plan buffer is full. And there are edge cases (or case?) when the previous line segment is currently beeing segmented and sent to the steppers?.
The best way to add this would be to split plan_buffer_line in two functions? Or make it reentrant (delay inserting the new line segment until after the calculation and arc insertion has been done) , calling it recursively?

> Or maybe even highjack the input gcode stream and add g2/3 arcs in between those straight lines and then let the normal bufferplanner do it's thing?

I do not think this is a good idea - better to let the planner handle it.

> Edit: maybe i can learn something from the LinuxCNC way of pathblending when i can find the right files in the repo.

[This](https://github.com/LinuxCNC/linuxcnc/blob/92df5135de183e61460adc21719bee4baaa57509/src/emc/tp/blendmath.c#L52) and [this](https://github.com/LinuxCNC/linuxcnc/blob/92df5135de183e61460adc21719bee4baaa57509/src/emc/tp/tp.c#L985) seems to be the two files involved?

> The NaiveCam stuff is black magic voodoo to me though

This, IMO, is the easy part - just extend the previous segment instead of adding a new one when the condition is met.

**Reply by Dietz0r** [reply]
> G187 should not be modified this way - a lookup table containing the factors should be used whether constant or not. Or a different gcode if not using the same syntax.

A lookup table sounds good! That can be done quickly and easily. I'll see to it. And apart from Fanuc with their G05 and the haas G187 it seems like any other MFG has their own way of calling those adjustments. Sticking to the Haas way seems ok to me, atleast it's documented properly. 

> All of them? IMO at least the code changes should be a compile time option initially. And G64 and G187 belongs to different modal groups (G13 and G0) creating a at least teoretical conflict.

It sounded a bit like "you either get path blending, or you get adjusted movement dynamics" that's what i disagreed with. 
If you are just talking about not calling them on the same line i'm on board. Although G187 is very much modal. It stays activated after calling until program end or disabled. (So it would be more at home in the G13 group than in the G0)

> Inserting the arc is not as easy as it seems, the calculation is at the end of adding a line segment and will most likely be done when the plan buffer is full. And there are edge cases (or case?) when the previous line segment is currently beeing segmented and sent to the steppers?. The best way to add this would be to split plan_buffer_line in two functions? Or make it reentrant (delay inserting the new line segment until after the calculation and arc insertion has been done) , calling it recursively?

Yeah ... I spent a bit of time yesterday looking at it and ran into the issue that you cant directly insert the arc anyways. 
Since the buffer is always fed mc_line lines. mc_arc generates a ton of those for arc movements and since we have limited buffer space we cant just shove those all in there either.  We should also not try to generate new arcs on the arc line segments we just added, otherwise we'll never finish, because there's always new intersections to calculate. 
Making the plan_buffer_line two functions could work. 🤔  one getting called from the mc_line (for the real line segment) and one from the end of the plan_buffer_line? (for the path blending)

> [This](https://github.com/LinuxCNC/linuxcnc/blob/92df5135de183e61460adc21719bee4baaa57509/src/emc/tp/blendmath.c#L52) and [this](https://github.com/LinuxCNC/linuxcnc/blob/92df5135de183e61460adc21719bee4baaa57509/src/emc/tp/tp.c#L985) seems to be the two files involved?

Thanks! 


PS: Have a good one and see you next year! :)

**Reply by Dietz0r** [reply]
I went ahead and added preliminary support for G187 and up to 5 Profiles. 
I kept in the non modal group because that seemed to be the easiest solution to not run into any conflicts like you suggested.
If you want to take a look [here's the dif](https://github.com/grblHAL/core/commit/4c28e420785c8e8e01b0e7b5fa19af3ba0a4681a) 

and of course i needed send a push right after that too with some variable type def and calculation cleanup: [Push here](https://github.com/grblHAL/core/commit/554812d5f13a27cc94c4d9ab5b03f1188b7b52e0) and [here.](https://github.com/grblHAL/core/commit/299a7cd1aa241aef9983e9cfd98348a2ed301745)

Hope that is in line with what you suggested? :)

**Reply by terjeio** [reply]
> I went ahead and added preliminary support for G187 and up to 5 Profiles.

The HAAS spec is only for 3? So a grblHAL specific code should be defined instead?

First, the new code should be encapsulated in #if #endif to make it an compile time option. There are at least two reasons for this. First is that the new code may introduce bugs/side-effects that may affect users that does not need the new functionality. Second is that the new setting(s) will triggers a settings reset that will affect everybody that updates whether they need the new functionality or not.
I am also still of the opinion that G187 and G64 should not normally be enabled simultaneously in the code but _could_ be made into a compile time option. This could be handled by one define in the same way I have done for the [Modbus implementation](https://github.com/grblHAL/core/blob/57243a0fc201fe4fdc97ad5880711a19eea49c0d/driver_opts.h#L329C1-L344C7) by [checking for bits](https://github.com/grblHAL/Plugins_spindle/blob/5a19330757c297af72b436dcf09ec4fdbf134143/modbus_rtu.c#L32) in the #define symbol.

The next issue to handle is how to handle the new parameters. As I understand it G64/G187 should only affect motions sent to the planner _after_ the commands are issued, not any that might already be in the planner. I have not studied the finer details of this but it could be that the relevant data has to be passed along to the planner instead of jerk and/or acceleration beeing overidden in settings when the commands are parsed.

**Reply by Dietz0r** [reply]
> The HAAS spec is only for 3? So a grblHAL specific code should be defined instead?

Yep, HAAS only has 3. If you want to change to something grblHAL specific, i'm totally with you. I mean i asked for a M-Code before. :)

> First, the new code should be encapsulated in #if #endif to make it an compile time option. There are at least two reasons for this. First is that the new code may introduce bugs/side-effects that may affect users that does not need the new functionality. Second is that the new setting(s) will triggers a settings reset that will affect everybody that updates whether they need the new functionality or not.

Will do, for both the Jerk and the Acceleration Profiles. 👍 

>I am also still of the opinion that G187 and G64 should not normally be enabled simultaneously in the code but could be made into a compile time option.

May I ask why? Maybe we're having different viewpoints or maybe something is unclear? I dont see the Problem with having both at the same time?

>The next issue to handle is how to handle the new parameters. As I understand it G64/G187 should only affect motions sent to the planner after the commands are issued, not any that might already be in the planner. I have not studied the finer details of this but it could be that the relevant data has to be passed along to the planner instead of jerk and/or acceleration beeing overidden in settings when the commands are parsed.

As far as i am aware both of the options are "Idle-State" only changes? G64 P? Q? switches the way the planner handles junctions so the planner should be empty (or purged and rebuild?) when changing parameters or path control mode? And G187 switches the whole machine dynamic. And while i could come up with scenarios where it would be handy to switch the acceleration profile mid operation, I believe that the most use it has is to define the dynamics before or inbetween individual operations.

>I have not studied the finer details of this but it could be that the relevant data has to be passed along to the planner instead of jerk and/or acceleration beeing overidden in settings when the commands are parsed.

I mean not changing the settings itself was the whole point of this thread. I could rig a gcode file that reads the needed $ entries and then modifies those too but it's not crash or usererror resistent. Or are you thinking about something else?

**Reply by terjeio** [reply]
>>  The HAAS spec is only for 3? So a grblHAL specific code should be defined instead?

> Yep, HAAS only has 3. If you want to change to something grblHAL specific, i'm totally with you. I mean i asked for a M-Code before. :)

If implementing G187 as HAAS compatible it should be according to that specification, if not another gcode should be selected, e.g. G327 or something like that. And it should be a G-code not a M-code since it will be in the core?

>> I am also still of the opinion that G187 and G64 should not normally be enabled simultaneously in the code but could be made into a compile time option.

> May I ask why? Maybe we're having different viewpoints or maybe something is unclear? I dont see the Problem with having both at the same time?

Because I do not like having both available at the same time unless explicitly asked for at compile time. Select none, G64, G187 or both.

>> The next issue to handle is how to handle the new parameters. As I understand it G64/G187 should only affect motions sent to the planner after the commands are issued, not any that might already be in the planner. I have not studied the finer details of this but it could be that the relevant data has to be passed along to the planner instead of jerk and/or acceleration beeing overidden in settings when the commands are parsed.

> As far as i am aware both of the options are "Idle-State" only changes? 

Ok, then either error out or wait until the planner is empty before temporarily changing the settings. 

>> I have not studied the finer details of this but it could be that the relevant data has to be passed along to the planner instead of jerk and/or acceleration beeing overidden in settings when the commands are parsed.

> I mean not changing the settings itself was the whole point of this thread. I could rig a gcode file that reads the needed $ entries and then modifies those too but it's not crash or usererror resistent. Or are you thinking about something else?

I was thinking of reading from the settings in the parser and passing modified values along to the planner in the plan structure, not modifying the underlying settings values. Then the planner would read the values off the plan structure not from the settings directly. The code will be more logical this way, at least as how I see it.
The settings override function was added so that a plugin could modify the acceleration.

**Reply by Dietz0r** [reply]
>If implementing G187 as HAAS compatible it should be according to that specification, if not another gcode should be selected, e.g. G327 or something like that. And it should be a G-code not a M-code since it will be in the core?
>Because I do not like having both available at the same time unless explicitly asked for at compile time. Select none, G64, G187 or both.

oh god... 
i think it just clicked why we're talking past each other.
You are talking about implementing the WHOLE HAAS routine ... G187 P-Value and E-Value and you are against having it alongside G64 because G64 the linuxcnc way does the same with P and Q values. I just wanted to implement the P-Values for the profiles and leave G64 alone. so i was sitting here all the time wondering why you'd be thinking pathblending and slower accelerations are a bad idea together.

for the record: i just want the profiles. i want to keep g64 like it is in linuxcnc and i dont want the g187 E-Values.
I'm with you in making it a grblHAL g-code if it is not exactly the same as HAAS's implementation so i'm ready to use whatever g or m-code you deem fit for those acceleration profiles and those alone.

> I was thinking of reading from the settings in the parser and passing modified values along to the planner in the plan structure, not modifying the underlying settings values. Then the planner would read the values off the plan structure not from the settings directly. The code will be more logical this way, at least as how I see it.
The settings override function was added so that a plugin could modify the acceleration.

That sounds like a very elegant solution. Basically the structure goes as this $Setting -> write into variable of planner during block generation with all applicable modifiers -> generate block? and then the acceleration profile would just alter the modifier for the planner block variable. Wouldnt this also clear the problem of the Buffer being full when you call the acceleration change since it then already has the correct values present before generating more blocks?

**Reply by Dietz0r** [reply]
I've tried implementing this for the acceleration profiles now.
Added a global value for an active profile and using the limit_acceleration function of the planner with a lookup to adjust the max allowable accelerations before returning the limit_value, thus reducing the values that the machine moves at without changing the underlying settings.

and the gcode simply switches the active profile value. 

Needs some testing now.
